File: string_calculator/go/src/xzr/string_calculator/string_calculator.go
package string_calculator

import (
	"container/list"
	"errors"
	"strconv"
	"strings"
)

func extractSeparator(numbers string) (sep string, nums string) {
	if numbers[0] == '/' {
		sepSpecEndIdx := strings.Index(numbers, "\n")
		return numbers[2:sepSpecEndIdx], numbers[sepSpecEndIdx+1:]
	}
	return ",", strings.Replace(numbers, "\n", ",", -1)
}

func toInts(numsAsStr []string) ([]int, error) {
	ints := make([]int, len(numsAsStr))
	for idx, n := range numsAsStr {
		num, e := strconv.Atoi(n)
		if e != nil {
			return ints, e
		}
		ints[idx] = num
	}
	return ints, nil
}

func partition(nums []int) (*list.List, *list.List) {
	negs := list.New()
	pos := list.New()
	for _, n := range nums {
		if n < 0 {
			negs.PushBack(n)
		} else {
			pos.PushBack(n)
		}
	}
	return negs, pos
}

func sum(nums *list.List) (sum int) {
	for it := nums.Front(); it != nil; it = it.Next() {
		sum += it.Value.(int)
	}
	return sum
}

func errorMsg(nums *list.List) (errMsg string) {
	errMsg = "negative numbers not allowed:"
	for it := nums.Front(); it != nil; it = it.Next() {
		errMsg += " " + strconv.Itoa(it.Value.(int))
	}
	return errMsg
}

func Add(numbers string) (int, error) {
	if len(numbers) == 0 {
		return 0, nil
	}
	sep, nums := extractSeparator(numbers)
	ints, e := toInts(strings.Split(nums, sep))
	if e != nil {
		return -1, e
	}
	negs, pos := partition(ints)
	if negs.Len() != 0 {
		return -1, errors.New(errorMsg(negs))
	} else {
		return sum(pos), nil
	}
}

File: string_calculator/elixir/string_calculator/lib/string_calculator.ex
defmodule StringCalculator do
	def add("") do
		0
	end

	def add(numsAsString) do
		numsAsString
		|> extractSepAndNums
		|> splitNumsBySep
		|> Enum.map(&(String.to_integer/1))
		|> Enum.partition(&(&1 < 0))
		|> sumPositives
	end

	defp extractSepAndNums("//" <> rest) do
		String.split(rest)
	end

	defp extractSepAndNums(numsAsString) do
		[",", String.replace(numsAsString, "\n", ",")]
	end

	defp splitNumsBySep([sep, nums]) do
		String.split(nums, sep)
	end

	defp sumPositives({[], pos}) do
		Enum.reduce(pos, &(+/2))
	end

	defp sumPositives({negs, _}) do
		negsAsString = Enum.map(negs, &(to_string/1)) |> Enum.reduce(&(&2 <> " " <> &1))
		raise(ArgumentError, message: "negative numbers ( " <> negsAsString <> " ) not allowed")
	end
end
